### PR TITLE
Add actual and upcoming overload : limitViolationInfos

### PR DESCRIPTION
**Pull request by souissimai**

### CODE REVIEW

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java [hunk 1]
```diff
@@ -143,6 +144,60 @@ private void cleanLoadFlowResultsAndPublishCancel(UUID resultUuid, String receiv
         notificationService.publishStop(resultUuid, receiver);
     }
 
+    private static LoadingLimits.TemporaryLimit handleEquipmentLimitViolation(Branch<?> branch, LimitViolationInfos violationInfo) {
+
+        Optional<com.powsybl.iidm.network.CurrentLimits> currentLimits = violationInfo.getSide().equals("ONE") ? branch.getCurrentLimits1() : branch.getCurrentLimits2();
+        if (violationInfo.getValue() < currentLimits.get().getPermanentLimit()) {
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java:150]:
```suggestion
        if (!currentLimits.isPresent() || violationInfo.getValue() < currentLimits.get().getPermanentLimit()) {
```

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java [hunk 2]
```diff
@@ -143,6 +144,60 @@ private void cleanLoadFlowResultsAndPublishCancel(UUID resultUuid, String receiv
         notificationService.publishStop(resultUuid, receiver);
     }
 
+    private static LoadingLimits.TemporaryLimit handleEquipmentLimitViolation(Branch<?> branch, LimitViolationInfos violationInfo) {
+
+        Optional<com.powsybl.iidm.network.CurrentLimits> currentLimits = violationInfo.getSide().equals("ONE") ? branch.getCurrentLimits1() : branch.getCurrentLimits2();
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java:149]:
```suggestion
        Optional<CurrentLimits> currentLimits = violationInfo.getSide().equals("ONE") ? branch.getCurrentLimits1() : branch.getCurrentLimits2();
```

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java [hunk 3]
```diff
@@ -143,6 +144,60 @@ private void cleanLoadFlowResultsAndPublishCancel(UUID resultUuid, String receiv
...
+        if (violationInfo.getValue() < currentLimits.get().getPermanentLimit()) {
+            return null;
+        } else {
+            List<LoadingLimits.TemporaryLimit> temporaryLimits = currentLimits.get().getTemporaryLimits().stream().collect(Collectors.toList());
+            Optional<LoadingLimits.TemporaryLimit> nextTemporaryLimit = temporaryLimits.stream()
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java:154]: there is no need for the tmp List
```suggestion
            Optional<LoadingLimits.TemporaryLimit> nextTemporaryLimit = currentLimits.get().getTemporaryLimits().stream()
```

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java [hunk 4]
```diff
@@ -143,6 +144,60 @@ private void cleanLoadFlowResultsAndPublishCancel(UUID resultUuid, String receiv
         notificationService.publishStop(resultUuid, receiver);
     }
 
+    private static LoadingLimits.TemporaryLimit handleEquipmentLimitViolation(Branch<?> branch, LimitViolationInfos violationInfo) {
```
Comment (dbraquart) [inline src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java:147]: - could we find a better describing name for this method ?   is it getBranchLowerLimitViolation() ? 
- can we add a comment somewhere to say:
              // limits are returned from the store by DESC duration / ASC value

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java [hunk 5]
```diff
@@ -143,6 +144,60 @@ private void cleanLoadFlowResultsAndPublishCancel(UUID resultUuid, String receiv
...
+            LoadingLimits.TemporaryLimit tempLimit = null;
+
+            Line line = network.getLine(equipmentId);
+            if (line != null) {
+                tempLimit = handleEquipmentLimitViolation(line, limitViolationInfo);
+            } else {
+                TwoWindingsTransformer twoWindingsTransformer = network.getTwoWindingsTransformer(equipmentId);
+                if (twoWindingsTransformer != null) {
+                    tempLimit = handleEquipmentLimitViolation(twoWindingsTransformer, limitViolationInfo);
+                }
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java:185]: I checked with PO and Techlead, we will not consider three windings transformer

File: src/main/java/org/gridsuite/loadflow/server/dto/LimitViolationInfos.java [hunk 6]
```diff
@@ -38,18 +40,21 @@ public class LimitViolationInfos {
...
             .subjectId(limitViolationEmbeddable.getSubjectId())
             .limit(limitViolationEmbeddable.getLimit())
             .limitName(limitViolationEmbeddable.getLimitName())
-            .acceptableDuration(limitViolationEmbeddable.getAcceptableDuration())
+            .actualOverload(limitViolationEmbeddable.getActualOverload())
+            .upComingOverload(limitViolationEmbeddable.getUpComingOverload())
             .value(limitViolationEmbeddable.getValue())
             .side(limitViolationEmbeddable.getSide())
             .limitType(limitViolationEmbeddable.getLimitType())
+            .actualOverload(limitViolationEmbeddable.getActualOverload())
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/dto/LimitViolationInfos.java:56]: done

File: src/main/java/org/gridsuite/loadflow/server/dto/LimitViolationInfos.java [hunk 7]
```diff
@@ -38,18 +40,21 @@ public class LimitViolationInfos {
...
+            .upComingOverload(limitViolationEmbeddable.getUpComingOverload())
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/dto/LimitViolationInfos.java:57]: done

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java (hunk 2 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java]: done

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java (hunk 3 above)
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java]: yes it's done

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java [hunk 8]
```diff
@@ -143,6 +144,60 @@ private void cleanLoadFlowResultsAndPublishCancel(UUID resultUuid, String receiv
...
+            }
+            return (tempLimit != null) ? tempLimit.getAcceptableDuration() : null;
+        }
+    }
+
+    private List<LimitViolationInfos> calculateOverloadLimitViolations(List<LimitViolationInfos> limitViolationInfos, Network network) {
+        for (LimitViolationInfos violationInfo : limitViolationInfos) {
+            if (violationInfo.getLimitName() != null && violationInfo.getLimitType().name().equals(CURRENT)) {
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java]: done

File: src/main/java/org/gridsuite/loadflow/server/dto/LimitViolationInfos.java [hunk 9]
```diff
@@ -29,7 +29,9 @@ public class LimitViolationInfos {
 
     private String limitName;
 
-    private Integer acceptableDuration;
+    private Integer actualOverload;
+
+    private Integer upComingOverload;
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/dto/LimitViolationInfos.java]: done

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java [hunk 10]
```diff
@@ -143,6 +144,60 @@ private void cleanLoadFlowResultsAndPublishCancel(UUID resultUuid, String receiv
         notificationService.publishStop(resultUuid, receiver);
     }
 
+    private static LoadingLimits.TemporaryLimit handleEquipmentLimitViolation(Branch<?> branch, LimitViolationInfos violationInfo) {
+
+        Optional<com.powsybl.iidm.network.CurrentLimits> currentLimits = violationInfo.getSide().equals("ONE") ? branch.getCurrentLimits1() : branch.getCurrentLimits2();
+        if (violationInfo.getValue() < currentLimits.get().getPermanentLimit()) {
+            return null;
+        } else {
+            List<LoadingLimits.TemporaryLimit> temporaryLimits = currentLimits.get().getTemporaryLimits().stream().collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java]: done, condition to check if currentLimits is present

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java]: yes I added condition to check if currentLimits is present

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java]: done

**Review by FranckLecuyer — APPROVED** [review]

**Review by dbraquart — CHANGES_REQUESTED** [review]
code: another rename requested
tests: ok

**remark:** Existing LF computation results cannot be displayed correctly using the new format. Did we plan to migrate the results (complex!), to remove all LF results (admin-tool script), to do nothing ?

File: src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java [hunk 11]
```diff
@@ -162,7 +214,8 @@ private void cancelLoadFlowAsync(LoadFlowCancelContext cancelContext) {
     public static LimitViolationInfos toLimitViolationInfos(LimitViolation violation) {
         return LimitViolationInfos.builder()
             .subjectId(violation.getSubjectId())
-            .acceptableDuration(violation.getAcceptableDuration())
+            .actualOverloadDuration(violation.getAcceptableDuration())
+            .upComingOverloadDuration(violation.getAcceptableDuration())
```
[the remark this answers is not in the document]
Comment (souissimai) [reply src/main/java/org/gridsuite/loadflow/server/service/LoadFlowWorkerService.java:218]: discussed

**Review by dbraquart — APPROVED** [review]
code: ok
tests: not done with last commit

**Pull request merged**